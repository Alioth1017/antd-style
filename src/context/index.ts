export * from './EmotionContext';
export * from './ThemeModeContext';
