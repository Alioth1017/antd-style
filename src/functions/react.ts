import _styled from '@emotion/styled';

import type { CreateStyled } from '@/types';

export const styled = _styled as CreateStyled;
export { css as reactCss, ThemeProvider as EmotionThemeProvider, withTheme } from '@emotion/react';
