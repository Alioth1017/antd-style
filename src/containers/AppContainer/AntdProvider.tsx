import { App, ConfigProvider } from 'antd';
import { ConfigProviderProps } from 'antd/es/config-provider';
import { type FC } from 'react';

import StaticMethod from './AntdStaticMethod';

const AntdProvider: FC<ConfigProviderProps> = ({ children, ...props }) => {
  return (
    <ConfigProvider {...props}>
      <App>
        <StaticMethod />
        {children}
      </App>
    </ConfigProvider>
  );
};

export default AntdProvider;
