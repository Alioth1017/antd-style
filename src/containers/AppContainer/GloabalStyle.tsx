import { createGlobalStyle, GlobalTheme } from '@/functions';
import { CSSStyle } from '@/types';
import { FC, memo } from 'react';

export interface GlobalStyleProps {
  /**
   * 全局样式
   */
  globalStyle?: CSSStyle<GlobalTheme>;
}

export const GlobalStyle: FC<GlobalStyleProps> = memo(({ globalStyle }) => {
  const Style = createGlobalStyle(...(globalStyle || []));

  return <Style />;
});
