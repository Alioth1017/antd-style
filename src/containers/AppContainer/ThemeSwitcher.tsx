import { FC, memo, ReactNode, useCallback, useLayoutEffect } from 'react';
import useControlledState from 'use-merge-value';

import { ThemeModeContext } from '@/context';
import { ThemeAppearance, ThemeMode } from '@/types';

let darkThemeMatch: MediaQueryList;

const matchThemeMode = (mode: ThemeAppearance) => matchMedia(`(prefers-color-scheme: ${mode})`);

export interface ThemeSwitcherProps {
  /**
   * 应用的展示外观主题，只存在亮色和暗色两种
   * @default light
   */
  appearance?: ThemeAppearance;
  defaultAppearance?: ThemeAppearance;
  onAppearanceChange?: (mode: ThemeAppearance) => void;
  /**
   * 主题的展示模式，有三种配置：跟随系统、亮色、暗色
   * 默认不开启自动模式，需要手动进行配置
   * @default light
   */
  themeMode?: ThemeMode;

  children: ReactNode;
}

const ThemeSwitcher: FC<ThemeSwitcherProps> = memo(
  ({
    children,
    appearance: appearanceProp,
    defaultAppearance,
    onAppearanceChange,
    themeMode = 'light',
  }) => {
    const [appearance, setAppearance] = useControlledState<ThemeAppearance>('light', {
      value: appearanceProp,
      defaultValue: defaultAppearance,
      onChange: onAppearanceChange,
    });

    const matchBrowserTheme = useCallback(() => {
      if (matchThemeMode('dark').matches) {
        setAppearance('dark');
      } else {
        setAppearance('light');
      }
    }, [setAppearance]);

    useLayoutEffect(() => {
      // 如果是自动的话，则去做一次匹配
      if (themeMode === 'auto') matchBrowserTheme();
      // 否则就明确设定亮暗色
      else setAppearance(themeMode);
    }, [themeMode]);

    useLayoutEffect(() => {
      if (!darkThemeMatch) {
        darkThemeMatch = matchThemeMode('dark');
      }

      if (!themeMode || themeMode === 'auto') {
        setTimeout(() => {
          matchBrowserTheme();
        }, 1);
      }

      darkThemeMatch.addEventListener('change', matchBrowserTheme);

      return () => {
        darkThemeMatch.removeEventListener('change', matchBrowserTheme);
      };
    }, []);

    return (
      <ThemeModeContext.Provider
        value={{
          themeMode,
          appearance,
          isDarkMode: appearance === 'dark',
        }}
      >
        {children}
      </ThemeModeContext.Provider>
    );
  },
);

export default ThemeSwitcher;
