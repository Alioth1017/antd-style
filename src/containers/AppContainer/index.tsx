import { App } from 'antd';
import { memo, ReactElement } from 'react';

import { ThemeProvider, ThemeProviderProps } from '../ThemeProvider';

export interface AppContainerProps<T, S = Record<string, string>> extends ThemeProviderProps<T, S> {
  className?: string;
}

export const AppContainer: <T, S>(props: AppContainerProps<T, S>) => ReactElement | null = memo(
  ({
    children,

    className,
    ...props
  }) => (
    <ThemeProvider {...props}>
      <App className={className}>{children}</App>
    </ThemeProvider>
  ),
);
