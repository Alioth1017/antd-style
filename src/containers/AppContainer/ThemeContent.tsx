import { GetAntdThemeConfig, GetCustomStylish, GetCustomToken } from '@/types';
import { ReactElement, ReactNode, useMemo } from 'react';

import { useThemeMode } from '@/hooks';
import { useAntdTheme } from '@/hooks/useAntdTheme';
import { theme } from 'antd';
import { ThemeConfig } from 'antd/es/config-provider/context';
import { ThemeProvider } from '../ThemeProvider';
import { AntdProvider, type AntdProviderProps } from './AntdProvider';

export interface ThemeContentProps<T, S = Record<string, string>>
  extends Omit<AntdProviderProps, 'theme'> {
  children: ReactNode;
  /**
   * 自定义 Token
   */
  customToken?: T | GetCustomToken<T>;
  /**
   * 自定义 Stylish
   */
  customStylish?: S | GetCustomStylish<S>;
  /**
   * 直接传入 antd 主题，或者传入一个函数，根据当前的主题模式返回对应的主题
   */
  theme?: ThemeConfig | GetAntdThemeConfig;
}

const ThemeContent: <T, S>(props: ThemeContentProps<T, S>) => ReactElement | null = ({
  children,
  customToken: customTokenOrFn,
  customStylish: stylishOrGetStylish,
  theme: themeProp,
  ...props
}) => {
  const { appearance, isDarkMode } = useThemeMode();
  const { stylish: antdStylish, ...token } = useAntdTheme();

  // 获取 自定义 token
  const customToken = useMemo(() => {
    if (customTokenOrFn instanceof Function) {
      return customTokenOrFn({ token, appearance, isDarkMode });
    }

    return customTokenOrFn;
  }, [customTokenOrFn, token, appearance]);

  // 获取 stylish
  const customStylish = useMemo(() => {
    if (stylishOrGetStylish instanceof Function) {
      return stylishOrGetStylish({
        token: { ...token, ...customToken },
        stylish: antdStylish,
        appearance,
        isDarkMode,
      });
    }
    return stylishOrGetStylish;
  }, [stylishOrGetStylish, token, customToken, antdStylish, appearance]);

  const antdTheme = useMemo<ThemeConfig>(() => {
    const baseAlgorithm = isDarkMode ? theme.darkAlgorithm : theme.defaultAlgorithm;

    let antdTheme = themeProp as ThemeConfig | undefined;

    if (typeof themeProp === 'function') {
      antdTheme = themeProp(appearance);
    }

    if (!antdTheme) {
      return { algorithm: baseAlgorithm };
    }

    // 如果有 themeProp 说明是外部传入的 theme，需要对算法做一个合并处理，因此先把 themeProp 的算法规整为一个数组
    const algoProp = !antdTheme.algorithm
      ? []
      : antdTheme.algorithm instanceof Array
      ? antdTheme.algorithm
      : [antdTheme.algorithm];

    return {
      ...antdTheme,
      algorithm: !antdTheme.algorithm ? baseAlgorithm : [baseAlgorithm, ...algoProp],
    };
  }, [themeProp, isDarkMode]);

  return (
    <AntdProvider theme={antdTheme} {...props}>
      <ThemeProvider customToken={customToken} customStylish={customStylish}>
        {children}
      </ThemeProvider>
    </AntdProvider>
  );
};

export default ThemeContent;
