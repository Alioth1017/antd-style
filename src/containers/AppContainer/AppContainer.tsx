import { memo, ReactNode } from 'react';

import { DisplayTheme, ThemeMode } from '@/types';

import { ThemeModeContext } from '@/context';
import { ThemeProvider, ThemeProviderProps } from '../ThemeProvider';
import { AntdProvider, type AntdProviderProps } from './AntdProvider';
import { GlobalStyle, GlobalStyleProps } from './GloabalStyle';

export interface AppContainerProps<T, S = Record<string, string>>
  extends ThemeProviderProps<T, S>,
    AntdProviderProps,
    GlobalStyleProps {
  /**
   * 应用的展示外观主题，只存在亮色和暗色两种
   * @default light
   */
  appearance?: DisplayTheme;
  defaultAppearance?: DisplayTheme;
  onAppearanceChange?: (mode: DisplayTheme) => void;
  /**
   * 主题的展示模式，有三种配置：跟随系统、亮色、暗色
   * 默认不开启自动模式，需要手动进行配置
   * @default light
   */
  themeMode?: ThemeMode;

  children: ReactNode;

  className?: string;
  prefixCls?: string;
}

export const AppContainer: <T, S>(props: AppContainerProps<T, S>) => ReactNode = memo(
  ({ children, appearance, themeMode, customToken, customStylish, globalStyle, ...props }) => {
    return (
      <ThemeModeContext.Provider
        value={{
          themeMode: themeMode || 'light',
          appearance: appearance || 'light',
        }}
      >
        <AntdProvider {...props}>
          <ThemeProvider customToken={customToken} customStylish={customStylish}>
            <GlobalStyle globalStyle={globalStyle} />
            {children}
          </ThemeProvider>
        </AntdProvider>
      </ThemeModeContext.Provider>
    );
  },
);
