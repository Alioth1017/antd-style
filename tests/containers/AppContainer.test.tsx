import { render, renderHook } from '@testing-library/react';
import { theme } from 'antd';
import { AppContainer, css, useTheme } from 'antd-style';
import { MappingAlgorithm } from 'antd/es/config-provider/context';
import { FC, PropsWithChildren } from 'react';
import { GetCustomToken } from '../../src/containers/AppContainer/ThemeContent';

declare module 'antd-style' {
  interface CustomToken {
    customColor: string;
    customBrandColor: string;
  }
  interface CustomStylish {
    x: string;
  }
}
const App: FC<{ id?: string }> = ({ id }) => {
  return (
    <div data-testid={id} style={{ padding: 16 }} className={'container'}>
      <a href="">节点样式</a>
    </div>
  );
};

describe('AppContainer', () => {
  it('默认', () => {
    const { container } = render(<App />, { wrapper: AppContainer });

    expect(container).toMatchSnapshot();
  });

  it('局部作用域', async () => {
    const { container, findByTestId } = render(
      <>
        <App id={'without'} />
        <AppContainer
          className={css`
            .container {
              color: red;
            }
          `}
        >
          <App id={'within'} />
        </AppContainer>
        ,
      </>,
    );
    expect(container).toMatchSnapshot();

    const nodeWithout = await findByTestId('without');
    const nodeWithin = await findByTestId('within');

    expect(nodeWithin).toHaveStyle('color: red;');
    expect(nodeWithout).not.toHaveStyle('color: red;');
  });

  describe('注入 antd 主题', () => {
    it('注入 token', () => {
      const Wrapper: FC<PropsWithChildren> = ({ children }) => (
        <AppContainer theme={{ token: { colorPrimary: '#000' } }}>{children}</AppContainer>
      );
      const { result } = renderHook(useTheme, { wrapper: Wrapper });

      expect(result.current.colorPrimary).toEqual('#000000');
    });

    it('仅注入算法', () => {
      const Wrapper: FC<PropsWithChildren> = ({ children }) => (
        <AppContainer theme={{ algorithm: theme.compactAlgorithm }}>{children}</AppContainer>
      );
      const { result } = renderHook(useTheme, { wrapper: Wrapper });

      expect(result.current.fontSize).toEqual(12);
    });

    it('注入暗色模式的自定义算法', () => {
      /**
       * 自定义主题算法
       * @param seedToken
       * @param mapToken
       */
      const customDarkAlgorithm: MappingAlgorithm = (seedToken, mapToken) => {
        const mergeToken = theme.darkAlgorithm(seedToken, mapToken);

        return {
          ...mergeToken,
          // Layout 颜色
          colorBgLayout: '#20252b',
          // 容器颜色
          colorBgContainer: '#282c34',
          // 悬浮类面板颜色
          colorBgElevated: '#32363e',
        };
      };

      const Wrapper: FC<PropsWithChildren> = ({ children }) => (
        <AppContainer
          appearance={'dark'}
          theme={(mode) =>
            mode === 'dark'
              ? { algorithm: [theme.compactAlgorithm, customDarkAlgorithm] }
              : undefined
          }
        >
          {children}
        </AppContainer>
      );
      const { result } = renderHook(useTheme, { wrapper: Wrapper });
      expect(result.current.colorPrimary).toEqual('#1668dc');
      expect(result.current.colorBgLayout).toEqual('#20252b');
      expect(result.current.fontSize).toEqual(12);
    });
  });

  describe('注入自定义主题', () => {
    it('注入自定义 Token', () => {
      const customTokenFn: GetCustomToken<any> = ({ token, isDarkMode }) => ({
        customColor: isDarkMode ? '#000' : token.colorPrimary,
        customBrandColor: isDarkMode ? token.colorPrimary : '#FFF',
      });

      const Wrapper: FC<PropsWithChildren> = ({ children }) => (
        <AppContainer customToken={customTokenFn}>{children}</AppContainer>
      );

      const { result: light } = renderHook(useTheme, { wrapper: Wrapper });
      expect(light.current.customColor).toEqual('#1677ff');
      expect(light.current.customBrandColor).toEqual('#FFF');

      const Darker: FC<PropsWithChildren> = ({ children }) => (
        <AppContainer appearance={'dark'} customToken={customTokenFn}>
          {children}
        </AppContainer>
      );

      const { result: dark } = renderHook(useTheme, { wrapper: Darker });
      expect(dark.current.customColor).toEqual('#000');
      expect(dark.current.customBrandColor).toEqual('#1677ff');
    });

    it('注入自定义 stylish', () => {
      const Wrapper: FC<PropsWithChildren> = ({ children }) => (
        <AppContainer customStylish={() => ({ x: '' })}>{children}</AppContainer>
      );

      const { result } = renderHook(useTheme, { wrapper: Wrapper });
      expect(result.current.stylish.x).toEqual('');
    });
  });

  describe('主题切换', () => {
    render(<AppContainer themeMode={'auto'}>App</AppContainer>);
  });
});
