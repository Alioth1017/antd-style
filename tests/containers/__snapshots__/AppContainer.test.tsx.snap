// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppContainer 局部作用域 1`] = `
.emotion-0 .container {
  color: red;
}

<div>
  <div
    class="container"
    data-testid="without"
    style="padding: 16px;"
  >
    <a
      href=""
    >
      节点样式
    </a>
  </div>
  <div
    class="ant-app emotion-0"
  >
    <div
      class="container"
      data-testid="within"
      style="padding: 16px;"
    >
      <a
        href=""
      >
        节点样式
      </a>
    </div>
  </div>
  ,
</div>
`;

exports[`AppContainer 默认 1`] = `
<div>
  <div
    class="ant-app"
  >
    <div
      class="container"
      style="padding: 16px;"
    >
      <a
        href=""
      >
        节点样式
      </a>
    </div>
  </div>
</div>
`;
